### ISSUE TITLE
`quotes` should allow template literals to avoid escaping single/double quotes

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#7660]
**What rule do you want to change?**

[`quotes`](http://eslint.org/docs/rules/quotes)

**Does this change cause the rule to produce more or fewer warnings?**

Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

This would change the default behavior of the `avoidEscape` option. Originally I thought the existing behavior was a bug, but based on the [option description](http://eslint.org/docs/rules/quotes#options) it appears to be working as documented.

**Please provide some example code that this change will affect:**

```js
/* eslint quotes: [error, "single", {"avoidEscape": true}] */

`this template literal contains both single and double quotes ' " `; // case 1

'this string contains both single and double quotes \' " '; // case 2

`this template literal does not have special characters`; // case 3

`this template literal only contains single quotes ' `; // case 4
```

**What does the rule currently do for this code?**

It reports 2 errors:

```
3:1  error  Strings must use singlequote  quotes // (case 1)
7:1  error  Strings must use singlequote  quotes // (case 3)
9:1  error  Strings must use singlequote  quotes // (case 4)
```

**What will the rule do after it's changed?**

It should only report an error for case 3 and 4. It should not report an error for case 1. Using something other than a template literal in this case would require escaping, which is undesirable because `avoidEscape: true` is set. However, it seems like a bad idea to report case 2, because (a) it would be backwards-incompatible, and (b) template literals are not supported in all environments. Case 4 should still be reported, because the rule would consider double quotes preferable to templates when avoiding escapes.

Note that an `allowTemplateLiterals` option already exists, but this option is more lenient than what I'm looking for; It allows *all* template literals, even if no quotes need to be escaped. (In the above example, ESLint reports no errors with the `allowTemplateLiterals: true` option.)

**Comment by kaicataldo** [comment]
Could we do an `ecmaVersion` check (defaulting to ES5) that would allow the rule to be smarter? There's precedence for this in the codebase: https://github.com/eslint/eslint/blob/master/lib/rules/func-name-matching.js#L93

I'm also assuming that this would also apply when the outer quotes are single or double. So the following should not warn with the enhancement, right?
```js
/* eslint quotes: [error, "single" or "double", {"avoidEscape": true}] */
`this template literal contains both single and double quotes ' " `;
```
```js
/* eslint quotes: [error, "backtick", {"avoidEscape": true}] */
'this template literal contains both backtick and double quotes ` " ';
"this template literal contains both backtick and single quotes ` ' ";
```

Strangely, for the "backtick" examples above, the first line currently errors while the second line does not. Don't have time to dig into the code at the moment, but curious why that is.

**Comment by platinumazure** [comment]
I'm not sure we should match against ecmaVersion because not all parsers would implement that (or would implement it using the same property name). I know we have precedent, but I'm not sure it's a good idea.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
